umbraco/Umbraco-CMS
# V8/feature/block list data structure v3

https://github.com/umbraco/Umbraco-CMS/pull/8339 ~**Must be reviewed + merged first!**~

Changed the code data structure of Block Editors to fit with data structure proposal v.3, see https://github.com/umbraco/rfcs/pull/23

This PR changes the data structure for both JS and c#. 

* ~Ensure all c# and JS tests pass~ ✔️ 

This also:

* Fixes/streamlines how the block editor data is serialized/deserialized
* Implements the ToEditor/FromEditor logic to ensure that these methods are called for all child controls
* Fixes some other minor issues found when testing 

## Testing

* If you already have a block editor and there's data in it, you won't get any errors, you're data will basically just be cleared because it will detect an invalid structure
* Ensure the block editor works as expected
* Ensure the multi text line editor works in the block editor

### Test case 1

This code should also deal with this [AB#7185](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/7185):

> If a user removes the a Block Type from Property Editor Configuration(Prevalues) I would like all the data based on that Block Type to be removed.
Be aware about that Block Editors can be nested inside each other.
> I do not think we need to re-publish data, as the item aren't rendered if configuration doesn't exists.
(Claus: correct - if we want to republish for any reason - we really need to consider unpublished data not getting published)

### Test case 2:

This code should also deal with this [AB#7187](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/7187):

> If an element type doesn't exist anymore we should not render the content of a block in the front-end(Website front-end). Same goes for content that does not match a block configuration from that property-editor.
